overlord/snapstate/handlers_components_mount_test.go

```diff
@@ -0,0 +1,294 @@
+	csi := snap.NewComponentSideInfo(cref, compRev)
+	t.Set("component-setup", snapstate.NewComponentSetup(csi, compPath))
+	t.Set("snap-setup", ssu)
+	chg := s.state.NewChange("test change", "change desc")
+	chg.AddTask(t)
+
+	s.state.Unlock()
+
+	s.se.Ensure()
+	s.se.Wait()
+
+	s.state.Lock()
+	c.Check(chg.Err(), IsNil)
+	s.state.Unlock()
+
+	// Ensure backend calls have happened with the expected data
+	c.Check(s.fakeBackend.ops, DeepEquals, fakeOps{
+		{
+			op: "setup-component",
+		},
+	})
+	// File not removed
+	c.Assert(osutil.FileExists(compPath), Equals, true)
+}
+
+func (s *mountCompSnapSuite) TestDoUndoMountComponent(c *C) {
+	const snapName = "mysnap"
+	const compName = "mycomp"
+	snapRev := snap.R(1)
+	compRev := snap.R(7)
+	ci, compPath := createTestComponent(c, snapName, compName)
+	si := createTestSnapInfoForComponent(c, snapName, snapRev, compName)
+	ssu := createTestSnapSetup(si, snapstate.Flags{})
+	s.AddCleanup(snapstate.MockReadComponentInfo(func(
+		compMntDir string) (*snap.ComponentInfo, error) {
+		return ci, nil
+	}))
+
+	s.state.Lock()
+	defer s.state.Unlock()
+
+	t := s.state.NewTask("mount-component", "task desc")
+	cref := naming.NewComponentRef(snapName, compName)
+	csi := snap.NewComponentSideInfo(cref, compRev)
+	t.Set("component-setup", snapstate.NewComponentSetup(csi, compPath))
+	t.Set("snap-setup", ssu)
+
+	chg := s.state.NewChange("sample", "...")
+	chg.AddTask(t)
+
+	terr := s.state.NewTask("error-trigger", "provoking total undo")
+	terr.WaitFor(t)
+	chg.AddTask(terr)
+
+	s.state.Unlock()
+
+	for i := 0; i < 3; i++ {
+		s.se.Ensure()
+		s.se.Wait()
+	}
+
+	s.state.Lock()
+
+	c.Check(chg.Err().Error(), Equals, "cannot perform the following tasks:\n"+
+		"- provoking total undo (error out)")
+
+	// ensure undo was called the right way
+	c.Check(s.fakeBackend.ops, DeepEquals, fakeOps{
+		{
+			op: "setup-component",
+		},
+		{
+			op: "undo-setup-component",
+		},
+		{
+			op: "remove-component-dir",
+		},
+	})
+}
+
+func (s *mountCompSnapSuite) TestDoMountComponentSetupFails(c *C) {
+	const snapName = "mysnap"
+	// fakeSnappyBackend.SetupComponent will fail for this component name
+	const compName = "broken"
+	snapRev := snap.R(1)
+	compRev := snap.R(7)
+	ci, compPath := createTestComponent(c, snapName, compName)
+	si := createTestSnapInfoForComponent(c, snapName, snapRev, compName)
+	ssu := createTestSnapSetup(si, snapstate.Flags{})
+	s.AddCleanup(snapstate.MockReadComponentInfo(func(
+		compMntDir string) (*snap.ComponentInfo, error) {
+		return ci, nil
+	}))
+
+	s.state.Lock()
+	defer s.state.Unlock()
+
+	t := s.state.NewTask("mount-component", "task desc")
+	cref := naming.NewComponentRef(snapName, compName)
+	csi := snap.NewComponentSideInfo(cref, compRev)
+	t.Set("component-setup", snapstate.NewComponentSetup(csi, compPath))
+	t.Set("snap-setup", ssu)
+
+	chg := s.state.NewChange("sample", "...")
+	chg.AddTask(t)
+
+	s.state.Unlock()
+
+	for i := 0; i < 3; i++ {
+		s.se.Ensure()
+		s.se.Wait()
+	}
```

Comment:
Why is the loop necessary? Shouldn't a single call work as well?

Reply:
The loop is needed when you have more than one task so more than one Ensure is needed. I've now restricted the loop to where it was necessary only.